### PR TITLE
[WIP] Numerical aquifer

### PR DESCRIPTION
**Pull request by GitPaean**
This is the first version of the implementation of numerical aquifers. 

The basic idea is to specify some cells as cells for numerical aquifers, and specify some connection to connect the cells to the reservoirs. 

For the numerical aquifer cells, we remove all the connections between them and other reservoir cells, then through NNCs to connection the cells to form numerical aquifers, also through NNCs we connect the aquifer cells to reservoir cells.

Some properties of the cells can be changed, like porosity, size, depth, permeability, pvtnum, satnum can be changed by the numerical aquifers. 

The results of the test case look good, but more tests will needed.

### CODE REVIEW

File: CMakeLists_files.cmake [hunk 1]
```diff
@@ -58,7 +58,10 @@ if(ENABLE_ECL_INPUT)
     src/opm/parser/eclipse/EclipseState/AquiferConfig.cpp
     src/opm/parser/eclipse/EclipseState/AquiferCT.cpp
     src/opm/parser/eclipse/EclipseState/Aquifetp.cpp
+    src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp
```
Comment (joakim-hove) [inline CMakeLists_files.cmake:61]: OK - when you are in this part of town. Can you make a `Aquifer/` directory and put all the aquifer stuff there.

File: opm/parser/eclipse/EclipseState/Aqucon.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#include <map>
+#include <array>
+
+namespace Opm {
+
+    class EclipseGrid;
+    class Deck;
+    class DeckRecord;
+
+    struct NumAquiferCon {
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Aqucon.hpp:35]: Can this share some code with the already existing implementation in `Aquancon`?

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+#include <iostream>
+
+#include "AquiferHelpers.hpp"
+
+namespace Opm {
+
+NumericalAquifers::NumericalAquifers(const Deck& deck, const EclipseGrid& grid, const FieldPropsManager& field_props)
+{
+    using AQUNUM=ParserKeywords::AQUNUM;
+    if ( !deck.hasKeyword<AQUNUM>() ) return;
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:42]: Each man his syntax :-)

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    // clear about the rules here. For now, we take care of all the keywords
+    const auto& aqunum_keywords = deck.getKeywordList<AQUNUM>();
+    for (const auto& keyword : aqunum_keywords) {
+        for (const auto& record : *keyword) {
+            const NumericalAquiferCell aqu_cell(record, grid, field_props);
+            // TODO: we should check whether the grid cell is active or NOT
+            // NOT sure what to do if the cell is inactive
+            if (this->hasCell(aqu_cell.global_index)) {
+                std::ostringstream ss;
+                ss << "NUMERICAL AQUIFER CELL AT GRID CELL { " << aqu_cell.I + 1
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:54]:
```C++
#include <opm/common/utility/OpmInputError.hpp>
#include <fmt/format.h>

auto error = fmt::format("Numerical aquifer cell at ({}, {}, {}) is declared more than once",
                                       aqu_cell.I + 1, aqu_cell.J + 1, aqu_cell.K + 1);
throw OpmInputError(error, keyword->location());

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+                   << " " << aqu_cell.J + 1 << " " << aqu_cell.K + 1 << " } IS DECLARED MORE THAN ONCE";
+                throw std::logic_error(ss.str());
+            } else {
+                this->addAquiferCell(aqu_cell);
+                auto& cells = this->aquifer_cells_;
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:59]: Would it not be natural to add the `aqufer_cells_` manipulation to the `addAquiferCell()` function?

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+
+        // For now, there is no two aquifers can be connected to one cell
+        // aquifer can not connect to aquifer cells
+        auto& aquifer = pair.second;
+        for (const auto& con : aqu_cons) {
+            const auto& aqu_con = con.second;
+            const size_t con_global_index = aqu_con.global_index;
+            if (this->hasCell(con_global_index)) {
+                const size_t cell_aquifer_id = this->getCell(con_global_index).aquifer_id;
+                std::ostringstream ss;
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:99]:
```
const auto& location = ???
auto msg = fmt::format("Problem with keyword {} \n"
                                     "In {} line {}\n"
                                     "Aquifer connecetion declared at grid cell () ....", 
                                     location.keyword, location.filename, location.lineno, 
                                     ......);
OpmLog::warning(msg);
```

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+NumericalAquiferCell::NumericalAquiferCell(const DeckRecord& record, const EclipseGrid& grid, const FieldPropsManager& field_props)
+   : aquifer_id( record.getItem<AQUNUM::AQUIFER_ID>().get<int>(0) )
+   , I ( record.getItem<AQUNUM::I>().get<int>(0) - 1 )
+   , J ( record.getItem<AQUNUM::J>().get<int>(0) - 1 )
+   , K ( record.getItem<AQUNUM::K>().get<int>(0) - 1 )
+   , area (record.getItem<AQUNUM::CROSS_SECTION>().getSIDouble(0) )
+   , length ( record.getItem<AQUNUM::LENGTH>().getSIDouble(0) )
+   , permeability( record.getItem<AQUNUM::PERM>().getSIDouble(0) )
+{
+    const auto& cell_depth = field_props.cellDepth();
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:177]: I would rather fetch the depth from the grid.

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+
+    if ( !record.getItem<AQUNUM::DEPTH>().defaultApplied(0) ) {
+        this->depth = record.getItem<AQUNUM::DEPTH>().getSIDouble(0);
+    } else {
+        this->depth = cell_depth[global_index];
+    }
+
+    if ( !record.getItem<AQUNUM::INITIAL_PRESSURE>().defaultApplied(0) ) {
+        this->init_pressure = record.getItem<AQUNUM::INITIAL_PRESSURE>().getSIDouble(0);
+    }
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:198]:
```C++
else
     this->init_pressure = number_of_the_beast??
```

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+        this->pvttable = pvtnum[global_index];
+    }
+
+    if ( !record.getItem<AQUNUM::SAT_TABLE_NUM>().defaultApplied(0) ) {
+        this->sattable = record.getItem<AQUNUM::SAT_TABLE_NUM>().get<int>(0);
+    } else {
+        this->sattable = satnum[global_index];
+    }
+
+    this->pore_volume = this->length * this->area * this->porosity;
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:212]: I guess this is on purpose - but you can ask the grid to get a proper cell volume(s)?

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp (hunk 8 above)
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:198]: You are right. I used to have it then it got removed when cleaning up.

File: src/opm/parser/eclipse/EclipseState/Aqucon.cpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                    const size_t aqu_id = con.aquifer_id;
+                    const size_t global_index = con.global_index;
+                    auto& aqu_cons = this->connections_[aqu_id];
+                    // TODO: with the following code, we might ignore the situation that a cell is
+                    // connected to two different aquifers
+                    if (aqu_cons.find(global_index) == aqu_cons.end()) {
+                        aqu_cons.insert({global_index, con});
+                    } else {
+                        std::ostringstream ss;
+                        ss << " the cell {" << con.I + 1 << " " << con.J + 1 << " " << con.K + 1 << " } is declared more "
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Aqucon.cpp:54]: `OpmInputError`

File: opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp [hunk 11]
```diff
@@ -42,10 +43,14 @@ class FieldPropsManager {
     FieldPropsManager() = default;
     FieldPropsManager(const Deck& deck, const Phases& ph, const EclipseGrid& grid, const TableManager& tables);
     virtual void reset_actnum(const std::vector<int>& actnum);
+    virtual void applyNumericalAquifer(const AquiferConfig&);
     const std::string& default_region() const;
     virtual std::vector<int> actnum() const;
     virtual std::vector<double> porv(bool global = false) const;
 
+    // TODO: golbal?
+    const std::vector<double>& cellDepth() const;
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp:52]: Do you *really* need this?

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp (hunk 9 above)
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:212]: No. It is this one to overwrite the grid one.

File: opm/parser/eclipse/EclipseState/NumericalAquifer.hpp [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    private:
+        // Maybe this id_ is not necessary
+        // Because if it is a map, the id will be there
+        // Then adding aquifer cells will be much easier with the
+        // default constructor
+        size_t id_;
+        std::vector<NumericalAquiferCell> cells_;
+        std::vector<NumAquiferCon> connections_;
+    };
+
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/NumericalAquifer.hpp:80]: Can you have the `NumericalAquifers` class in a separate .cpp / .hpp file

File: opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp (hunk 11 above)
Comment (GitPaean) [reply opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp:52]: To be honest, NO. Because the cell_depth in `FieldProps` is NOT used for anything in `Flow`, but it is something found later. 

In the simulator code, the depth is calculated based on averaging of the corner points.

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+                std::ostringstream ss;
+                ss << "AQUIFER CONNECTION DECLARD AT GRID CELL { " << aqu_con.I + 1 << " " << aqu_con.J + 1 << " "
+                    << aqu_con.K + 1 << " } IS A AQUIFER CELL OF AQUIFER " << cell_aquifer_id << ", AND WILL BE REMOVED";
+                OpmLog::warning(ss.str());
+                continue;
+            }
+
+            if (con_set.find(con_global_index) != con_set.end()) {
+                // TODO: not totally sure whether a cell can be connected two different aquifers
+                std::ostringstream ss;
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:108]: `OpmInputError`

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp (hunk 7 above)
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:177]: `Grid` does not have it . It is always using the corner depth to do the average. Maybe we can modify the `m_coord` to achieve the goal. But I am not familiar enough with grid to give a judgement. 

Anyway, I will remove the update of `cell_depth` here.

File: opm/parser/eclipse/EclipseState/NumericalAquifer.hpp [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        double transmissibility;
+        size_t global_index;
+    };
+
+    class SingleNumericalAquifer {
+    public:
+        explicit SingleNumericalAquifer(const size_t aqu_id);
+        void addAquiferCell(const NumericalAquiferCell& aqu_cell);
+        void addAquiferConnection(const NumAquiferCon& aqu_con);
+        void updateCellProps(const EclipseGrid& grid,
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/NumericalAquifer.hpp:63]: The way I understand it the `updateCellProps` method should be called by the simulator to update the `pore_volume, satnum, pvtnum` and `cell_depth` arrays? Allowing the opm-common aquifer implementation to overwrite the core data in this way is in my opinion quite an invasion of privacy. I think you should rather export a vector of small structs and then let the simulator tinker with it's own data; i.e. something like:
```
// opm-common:

struct AquiferCell {
      std::size_t global_index;
      double pore_volume; 
      int satnum;
      int pvtnum;
      double cell_depth;
}
 
std::vector<AquiferCell> exportCellProps(const EclipseGrid& grid) {
}

// simulator
auto aquifer_cells = aquifer_object.exportCellProps(grid);

for (const auto& aq_cell : aquifer_cells) {
     ....
}
```

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp (hunk 8 above)
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:198]: A question here, is using some kind of `NaN` a good practice?  `Beast number` might pollute the results in a way not easy to find. Like 
https://github.com/OPM/opm-models/issues/635
The fix is from the opm-grid
https://github.com/OPM/opm-grid/pull/498

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+    const auto& poro = field_props.get_double("PORO");
+    const auto& pvtnum = field_props.get_int("PVTNUM");
+    const auto& satnum = field_props.get_int("SATNUM");
+
+    this->global_index = grid.getGlobalIndex(I, J, K);
+
+    if ( !record.getItem<AQUNUM::PORO>().defaultApplied(0) ) {
+        this->porosity = record.getItem<AQUNUM::PORO>().getSIDouble(0);
+    } else {
+        this->porosity = poro[global_index];
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:187]: The vectors from `field_props` only have active elements; i.e. you must use `active_index` when looking up values here.

File: src/opm/parser/eclipse/share/keywords/000_Eclipse100/A/AQUCON [hunk 16]
```diff
@@ -39,27 +39,27 @@
...
       "value_type": "DOUBLE",
-      "default_value": 1
+      "default": 1
     },
     {
       "name": "TRANS_OPTION",
       "value_type": "INT",
-      "default_value": 0
+      "default": 0
     },
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/share/keywords/000_Eclipse100/A/AQUCON:48]: If you pull this change out in a separate PR we will merge it right away!

File: opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp (hunk 11 above)
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp:52]: > In the simulator code, the depth is calculated based on averaging of the corner points.

Also this value is based on the average of the corner points (in the EclipseGrid) - it is just cached here for faster initiialization of the saturation properties.

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp (hunk 15 above)
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:187]: You are right.

File: opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp (hunk 11 above)
Comment (GitPaean) [reply opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp:52]: Are you sure it is used somewhere? I did not find it.

**Comment by joakim-hove** [comment]
OK - I'll stop reviewing now, please ask/complain if you have objections to my points. If possible I would encourage you to extract small & trivial parts from this and make separate PR's for that?

File: opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp (hunk 11 above)
Comment (GitPaean) [reply opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp:52]: Okay. It looks like it is used in ```SatfuncPropertyInitializers.cpp`.

**Comment by joakim-hove** [comment]
> The results of the test case look good, but more tests will needed.

And you my friend need to add some!

**Comment by GitPaean** [comment]
> OK - I'll stop reviewing now, please ask/complain if you have objections to my points. If possible I would encourage you to extract small & trivial parts from this and make separate PR's for that?

Thanks for the input.

> And you my friend need to add some!

Tests will be there.

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp (hunk 8 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:201]: > A question here, is using some kind of NaN a good practice? 

Thank you for asking that question - let me introduce my good friend for situations like that: `std::optional<double>`

File: opm/parser/eclipse/EclipseState/NumericalAquifer.hpp (hunk 14 above)
Comment (GitPaean) [reply opm/parser/eclipse/EclipseState/NumericalAquifer.hpp:63]: I understand `updateCellProps` is invasive.  The main way I hope this functionality works is that we fix all (as much as possible) the information when we parsing it, then at the simulator side, we do not need to do much or nothing. 

From the parser to the simulator, we have a lot of good implementation and design there, hopefully the functionality can work  seamless with all these good implementation and design. 

If we check the `aquifer_cells` whenever we touch the aquifer cells, there might be a lot of places we need to do it and it will be invasive in another manner and might not be easy to get things correct. 

There is some challenges here is that in the simulator code might use different ways to obtain information. For example, `cell_depth` is not used in the simulator code, which makes that we have to check the `aquifer_cells` to get the depth information.

**Comment by GitPaean** [comment]
New finding is that AQUNUM can activate the grid cells deactivated through `ACTNUM`. The PR will adapted to address that.

**Comment by joakim-hove** [comment]
> New finding is that AQUNUM can activate the grid cells deactivated through ACTNUM. The PR will adapted to address that.

Now that is a quite **major change**. If possible I would like to isolate that change to one (small) PR.

File: opm/parser/eclipse/EclipseState/NumericalAquifer.hpp (hunk 14 above)
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/NumericalAquifer.hpp:63]: OK - we will take our time and find a good solution together. Thank you for opening the PR early.

**Comment by GitPaean** [comment]
> Now that is a quite major change. If possible I would like to isolate that change to one (small) PR.

A smaller and more general/proper PR will be to use `PORV` to activate cells that are deactivated by `ACTNUM`. If I understand correctly, in the current implementation of OPM, in the `FieldProps`, all the information of the cells deactivated by `ACTNUM` are discarded already. Through `FieldProps`, it can only deactivate more cells while you can not activate inactive cells. 

At least for now, it looks to me it can be pretty involved to support this which will include a major refactoring of the way how `FieldProps` works. Due to the urgency of getting the correct results, I might not be able to prioritize this. 

I have to go ahead to make sure the targeting case can obtain correct results first, which is really important now.

Anyone familiar with implementation of the `FieldProps`, probably @joakim-hove the best, if having the capacity and willing to do it, can go ahead to refactor this to support using `PORV` to activate inactive cells.

**Comment by joakim-hove** [comment]
If you are now struggling with summary output I suggest we merge what works now. Remember: the time to complere a task increases styringer than linearly with rask side - divide and conquer!

**Comment by GitPaean** [comment]
> If you are now struggling with summary output I suggest we merge what works now. 

Thanks for the suggestion, the summary output looks half good now. There is something wrong with the flux collection for summary output, I hope to finish it very soon.

**Comment by GitPaean** [comment]
The functionality development look okay unless some other tests show otherwise. I will begin working on cleaning up and getting the PR ready for merge.

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp (hunk 3 above)
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:42]: This is the only situation that I do not add `{` for `if` statement.

File: CMakeLists_files.cmake (hunk 1 above)
Comment (GitPaean) [reply CMakeLists_files.cmake:61]: Will do it.

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp (hunk 3 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:42]: I think we have found a sort of truce when it comes to syntax; I just commented on the `return;` on the same line. It is fine - leave it!

**Comment by joakim-hove** [comment]
I really think you should Try to merge this om pieces.

**Comment by GitPaean** [comment]
Another finding, we need to begin counting the aquifer connections after we finish all the grid processing. 

Because a cell can only connect to an aquifer through boundary faces (by default), the grid processing will change the reservoir boundary, so the definition of the boundary faces will be different after the grid processing.  

Basically, we will not be able to count the connections in opm-common.

**Comment by GitPaean** [comment]
Now the PR  (along with https://github.com/OPM/opm-simulators/pull/2910 and https://github.com/OPM/opm-grid/pull/502 ) is pure prototyping to get the correct results for testing. Will create new PR to finalize the work.

**Comment by GitPaean** [comment]
jenkins build this opm-grid=502 opm-simulators=2910 please

**Comment by GitPaean** [comment]
jenkins build this opm-grid=502 opm-simulators=2910 please

**Comment by GitPaean** [comment]
The testing results for the related PRs (OPM/opm-simulators#2910 and OPM/opm-grid#502 ) in the field case are satisfactory.  We will cleaning these PR now.

**Comment by GitPaean** [comment]
> I really think you should Try to merge this om pieces.

Sure. I will do that. There will be different PRs to get this work in.

File: src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp [hunk 17]
```diff
@@ -0,0 +1,419 @@
...
+    std::cout << " fp.active_size " << fp.active_size()  << std::endl;
+    std::cout << " grid.grid.getNumActive() " << grid.getNumActive() << std::endl;
+    const auto& cell1 = this->cells_[0];
+    // all the connections connect to the first numerical aquifer cell
+    const size_t gc1 = cell1.global_index;
+    for (const auto& con : this->connections_) {
+        const size_t gc2 = con.global_index;
+        // TODO: the following only works for Cartesian grids, more tests need to done
+        // for more general grid
+        const auto& cell_dims = grid.getCellDims(gc2);
```
Comment (GitPaean) [inline src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp:336]: Not totally sure whether we should use the EclipseGrid here.
